src/hardware/mouse/mouse_queue.cpp

```diff
@@ -0,0 +1,317 @@
+	}
+
+	// Check if we still need this event
+	if (!ev.dos_moved && !ev.dos_wheel && !ev.dos_button)
+		ev.request_dos = false;
+}
+
+void MouseQueue::FetchEvent(MouseEvent &ev)
+{
+	// First try (prioritized) DOS events
+	if (HasReadyEventDos()) {
+		DEBUG_QUEUE("FetchEvent %s", "DOS");
+		// Mark event as DOS one
+		ev.request_dos = true;
+		ev.dos_moved   = pending_dos_moved;
+		ev.dos_button  = pending_dos_button;
+		ev.dos_wheel   = pending_dos_wheel;
+		ev.dos_buttons = pending_dos_buttons_state;
+		// Set delay before next DOS events
+		delay.dos_ms = start_delay.dos_ms;
+		// Clear event information
+		pending_dos_moved  = false;
+		pending_dos_button = false;
+		pending_dos_wheel  = false;
+		return;
+	}
+
+	// Now try PS/2 event
+	if (HasReadyEventPS2()) {
+		DEBUG_QUEUE("FetchEvent %s", "PS2");
+		// Set delay before next PS/2 events
+		delay.ps2_ms = start_delay.ps2_ms;
+		// PS/2 events are really dummy - merely a notification
+		// that something has happened and driver has to react
+		ev.request_ps2 = true;
+		pending_ps2    = false;
+		return;
+	}
+
+	// Nothing to provide to interrupt handler,
+	// event will stay empty
+}
+
+void MouseQueue::ClearEventsDOS()
+{
+	// Clear DOS relevant part of the queue
+	pending_dos_moved  = false;
+	pending_dos_button = false;
+	pending_dos_wheel  = false;
+	delay.dos_ms       = 0;
+
+	// If timer is not needed, stop it
+	if (!HasEventAny()) {
+		timer_in_progress = false;
+		PIC_RemoveEvents(mouse_queue_tick);
+	}
+}
+
+void MouseQueue::StartTimerIfNeeded()
+{
+	// Do nothing if timer is already in progress
+	if (timer_in_progress)
+		return;
+
+	bool timer_needed = false;
+	uint8_t delay_ms  = UINT8_MAX; // dummy delay, will never be used
+
+	if (HasEventPS2() || delay.ps2_ms) {
+		timer_needed = true;
+		delay_ms     = std::min(delay_ms, delay.ps2_ms);
+	}
+	if (HasEventDos() || delay.dos_ms) {
+		timer_needed = true;
+		delay_ms     = std::min(delay_ms, delay.dos_ms);
+	}
+
+	// If queue is empty and all expired, we need no timer
+	if (!timer_needed)
+		return;
+
+	// Enforce some non-zero delay between events; needed
+	// for example if DOS interrupt handler is busy
+	delay_ms = std::max(delay_ms, static_cast<uint8_t>(1));
+
+	// Start the timer
+	DEBUG_QUEUE("StartTimer, %d", delay_ms);
+	pic_ticks_start   = PIC_Ticks;
+	timer_in_progress = true;
+	PIC_AddEvent(mouse_queue_tick, static_cast<double>(delay_ms));
+}
+
+void MouseQueue::UpdateDelayCounters()
+{
+	const uint32_t tmp = (PIC_Ticks > pic_ticks_start)
+	                           ? (PIC_Ticks - pic_ticks_start)
+	                           : 1;
+	uint8_t elapsed    = static_cast<uint8_t>(
+                std::min(tmp, static_cast<uint32_t>(UINT8_MAX)));
+	if (!pic_ticks_start)
+		elapsed = 1;
+
+	auto calc_new_delay = [](const uint8_t delay, const uint8_t elapsed) {
+		return static_cast<uint8_t>((delay > elapsed) ? (delay - elapsed)
+		                                              : 0);
+	};
```

Comment:
`delay` is also a member of the `MouseQueue`; suggest a new local name to avoid ambiguity/shadowing.

Reply:
OK, changed argument name.